src/test/cs/Infinispan/HotRod/ContinuousQueryTest.cs

```diff
@@ -0,0 +1,383 @@
+    }
+    class ValueMarshaller : IMarshaller
+    {
+        public bool IsMarshallable(object o)
+        {
+            throw new NotImplementedException();
+        }
+
+        public object ObjectFromByteBuffer(byte[] buf)
+        {
+            return User.Parser.ParseFrom(buf);
+        }
+
+        public object ObjectFromByteBuffer(byte[] buf, int offset, int length)
+        {
+            throw new NotImplementedException();
+        }
+
+        public byte[] ObjectToByteBuffer(object obj)
+        {
+            throw new NotImplementedException();
+        }
+
+        public byte[] ObjectToByteBuffer(object obj, int estimatedSize)
+        {
+            throw new NotImplementedException();
+        }
+    }
+    class ContinuousQueryTest
+    {
+        RemoteCacheManager remoteManager;
+        const String ERRORS_KEY_SUFFIX = ".errors";
+        const String PROTOBUF_METADATA_CACHE_NAME = "___protobuf_metadata";
+        const String NAMED_CACHE = "InMemoryNonSharedIndex";
+
+        [TestFixtureSetUp]
+        public void BeforeClass()
+        {
+            ConfigurationBuilder conf = new ConfigurationBuilder();
+            conf.AddServer().Host("127.0.0.1").Port(11222).ProtocolVersion("2.6");
+            conf.ConnectionTimeout(90000).SocketTimeout(6000);
+            conf.Marshaller(new BasicTypesProtoStreamMarshaller());
+            remoteManager = new RemoteCacheManager(conf.Build(), true);
+
+            IRemoteCache<String, String> metadataCache = remoteManager.GetCache<String, String>(PROTOBUF_METADATA_CACHE_NAME);
+            metadataCache.Put("sample_bank_account/bank.proto", File.ReadAllText("proto2/bank.proto"));
+            if (metadataCache.ContainsKey(ERRORS_KEY_SUFFIX))
+            {
+                Console.WriteLine("fail: error in registering .proto model");
+                Environment.Exit(-1);
+            }
+
+            IRemoteCache<int, User> userCache = remoteManager.GetCache<int, User>(NAMED_CACHE);
+        }
+        [TestFixtureTearDown]
+        public void AfterClass()
+        {
+            remoteManager.Stop();
+        }
+        [Test]
+        public void EntityBasicContQueryTest()
+        {
+            int joined = 0, updated = 0, leaved = 0;
+            try
+            {
+                IRemoteCache<int, User> userCache = remoteManager.GetCache<int, User>(NAMED_CACHE);
+                Semaphore s = new Semaphore(0, 1);
+                QueryRequest qr = new QueryRequest();
+                // JpqlString will be deprecated please use QueryString
+                // qr.JpqlString = "from sample_bank_account.User";
+                qr.QueryString = "from sample_bank_account.User";
+
+                Func<object, User> vFromByteBuffer = (b) => new Google.Protobuf.MessageParser<User>(() => new User()).ParseFrom((byte[])b);
+                Func<object, int> kFromByteBuffer = (b) => new Google.Protobuf.MessageParser<base_types>(() => new base_types()).ParseFrom((byte[])b).I32;
```

Comment:
I'd say that these to functions are unnecessary and redundant in the API. IMO, it should be responsibility of the Marshaller to unmarshall data. This functionality should be part of the marshaller and the continous query engine should be able to use the marshaller according to the type parameters of the ContinuousQueryListener. This would be more user friendly, IMO.

Reply:
You can also look at RemoteQueryUtils in the C# client test suite. There's some mechanism for detecting the message parser automatically from the type. Maybe this could be used here.